### ISSUE TITLE
Support managing @types along with their counter-part dependency

### ISSUE THREAD

**Issue by aaronbeall** [npm/npm#19315]
#### What's the feature?

When using TypeScript you often need to install a package along with a type definition, as in `npm install react @types/react`. The feature request is to have `npm` aware of this relationship in a TypeScript project, roughly like:
- If an installed/updated package has built-in type-definitions, there's no change.
- Otherwise, if a `@types/<package>` exists, install/update with the same version.
- If a `@types/<package>` does not exist warn the user.

A TypeScript project could be figured from package.json or a separate CLI argument, like `npm install foo -ts`.

Perhaps this is a good use-case for an external tool, but it would be awesome if `npm` had built in awareness of this.

#### What problem is the feature intended to solve?

When installing, updating, choosing or changing versions, or uninstalling you have to manually keep these two things in sync. It's easy to set one and not the other, and get false or confusing errors, unexpected runtime errors, or lack of correct type checking. It's also a bit manual to figure out if a package provides its own type-definitions or if a separate `@types` package is needed, to the point that some popular libraries provide "dummy" `@types` packages just to avoid an install error.

#### Is the absence of this feature blocking you or your team? If so, how?

No.

#### Is this feature similar to an existing feature in another tool?

Probably?

#### Is this a feature you're prepared to implement, with support from the npm CLI team?

Yes.

**Comment by kwonoj** [comment]
I am in great favor if npm cli can support this kind of sceanarios, in addition to above cases also support one another cases as well
- if package have type definition but there are `@types` are installed, warn / or remove `@types` package
: This usually happens some package introduces own type definition later time while there are 3rd party types were existing, some type pkgs are aware of it and provide warning message but some are not, so user manually have to verify newly updated pkg have own type definition to remove 3rd party types.

**Comment by kwonoj** [comment]
🤔  meanwhile second thought: what'll happen cases of corresponding `@types` package have bug for type definition, so intended to not upgrade? I have few modules for those kind of cases actually.

**Comment by aaronbeall** [comment]
> what'll happen cases of corresponding @types package have bug for type definition, so intended to not upgrade? 

Good point, this definitely would need to be fully fleshed out for all use cases. My general thinking with versioning (based on how I've handled various "odd" cases of type def versions manually) is that the versioning for `@types` would automatically update as long as the semantic version is compatible with the correlated package, but if you manually break the version compatibility it will no longer make changes to the `@types` package automatically. In other words if at install/update time the version in `@types` is not semantically flexible in the same way as the dependency, it will leave it as is, and maybe issue a warning about a possible mismatch.

To play that out for say React, today if you `npm install react` you get `"react": "^16.2.0"`, the feature I'm proposing here would also install `"@types/react": "^16.0.27"` (as if you did `npm install @types/react` manually). Due to the `^16` used in `@types/react` we can say its compatible with `react^16` and we're good. But where it gets interesting is if React releases version 17 before the type defs, or type defs release a breaking type-def we *don't* want like you said:
- If React 17 is released and we install it, and there's not yet a type-defs beyond 16, we get a warning. This is not an ideal situation and the user should be aware of the mismatch, but its possible the type-defs are ok or "good enough" even at a major version behind. In the case of react type-defs tend to update with the version, but other less popular libraries can have the type-defs get multiple versions behind... user really needs to be aware of this to identify out-of-date type-defs, but not stopped completely.
- If `@types/react` releases version `16.0.28` that has a bug and breaks things (of course it *shouldn't*, but it happens), we can manually install the exact version or range before it broke, like `npm install @types/react@16.0.27`. Now when I install or update `react`, I'll get a warning that the version of `@types` is *behind* the version of `react`, but no automatic install/update of `@types/react` is going to happen at this point. 

I'm probably way over-simplifying this and missing a lot of other cases. :) But I think this is a problem worth solving. 

Obviously, good messaging is hard, but will go a long ways. This is really all about user awareness of how their type-defs and correlated packages are staying in sync... which today is quite difficult to manage in large projects.
